text/0000-trait-alias.md

```diff
@@ -0,0 +1,266 @@
+
+fn iterate<SI>(si: SI) where SI: StringIterator {} // used as bound
+```
+
+#### As a trait object
+
+```rust
+fn iterate_object(si: &StringIterator) {} // used as trait object
+```
+
+#### In an `impl Trait`
+
+```rust
+fn string_iterator_debug() -> impl Debug + StringIterator {} // used in an impl Trait
+```
+
+As shown above, a *trait alias* can substitute associated types. It doesn’t have to substitute them all. In that case, the *trait alias*
+is left incomplete and you have to pass it the associated types that are left. Example with the
+[tokio case]([tokio example](#second-example-ergonomic-collections-and-scrapping-boilerplate):
+
+```rust
+pub trait Service {
+  type Request;
+  type Response;
+  type Error;
+  type Future: Future<Item=Self::Response, Error=Self::Error>;
+  fn call(&self, req: Self::Request) -> Self::Future;
+}
+
+trait HttpService = Service<Request = http::Request, Response = http::Response, Error = http::Error>;
+
+trait MyHttpService = HttpService<Future = MyFuture>; // assume MyFuture exists and fulfills the rules to be used in here
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Adds another construct to the language.
+
+- The syntax `trait TraitAlias = Trait` requires lookahead in the parser to disambiguate a trait from a trait alias.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- It’s possible to create a new trait that derives the trait to alias, and provide a universal `impl`:
+    
+    ```rust
+    trait Foo {}
+    
+    trait FooFakeAlias: Foo {}
+    
+    impl<T> Foo for T where T: FooFakeAlias {}
+    ```
+    
+    This works for trait objects and trait bounds only. You cannot implement `FooFakeAlias` directly
+    because you need to implement `Foo` first – hence, you don’t really need `FooFakeAlias` if you can
+    implement `Foo`.
+    
+    There’s currently no alternative to the impl problem described here.
+
+- Similar to Haskell's ContraintKinds, we could declare an entire predicate as a reified list of constraints, instead of creating an alias for a set of supertraits and predicates. Syntax would be something like `constraint Foo<T> = T: Bar, Vec<T>: Baz;`, used as `fn quux<T>(...) where Foo<T> { ... }` (i.e. direct substitution). Trait object usage is unclear.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- Should we use `type` as the keyword instead of `trait`?
+
+    `type Foo = Bar;` already creates an alias `Foo` that can be used as a trait object.
+    
+    If we used `type` for the keyword, this would imply that `Foo` could also be used as a bound as well. If we use `trait` as proposed in the body of the RFC, then `type Foo = Bar;` and `trait Foo = Bar;` _both_ create an alias for the object type, but only the latter creates an alias that can be used as a bound, which is a confusing bit of redundancy.
+    
+    However, this mixes the concepts of types and traits, which are different, and allows nonsense like `type Foo = Rc<i32> + f32;` to parse.
```

Comment:
I actually...would expect that to parse, but fail in semantical resolution, since `Rc<i32>` and `f32` are not traits. That is, something like `type Foo = Write + Send;` ought to be legal today, I would think, and has a very similar structure. 

It's probably worth noting that this also interacts with the current way that we "pun" on a trait-name for the object type. If we adopted some keyword for objects, like `obj Trait`, then `type Foo = obj (Write + Send)` wouldn't make much sense for use in a where-clause (but one could have `type Foo = Write + Send` play some sort of "dual-role", becoming something that can be used only in an `obj Foo` type but not as a type itself -- this is, of course, *preecisely* the role that `trait Foo = Write + Send` would play under the current RFC)